design/20210209.certificates.k8s.io-adoption.md

```diff
@@ -0,0 +1,292 @@
+---
+title: certificates.k8s.io Adoption
+authors:
+  - "@joshvanl"
+reviewers:
+  - "@joshvanl"
+approvers:
+  - "@munnerz"
+  - @jetstack/team-cert-manager
+editor: "@joshvanl"
+creation-date: 2021-02-09
+last-updated: 2021-02-17
+status: provisional
+---
+
+# certificates.k8s.io Adoption
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+- [Reasons for Approach](#reasons-for-approach)
+  * [Read all access to `CertificateSigningRequests`](#read-all-access-to-certificatesigningrequests)
+  * [Namespaced issuers referenced by `CertificateSigningRequest`](#namespaced-issuers-referenced-by-certificatesigningrequest)
+  * [Duration Field](#duration-field)
+  * [CA Field](#ca-field)
+  * [External Issuers](#external-issuers)
+- [Proposal](#proposal)
+  * [Signers](#signers)
+  * [Conditions](#conditions)
+  * [API Changes](#api-changes)
+  * [Upgrading](#upgrading)
+  * [Graduation](#graduation)
+<!-- /toc -->
+
+## Summary
+
+In Kubernetes v1.19 the
+[`CertificateSigningRequest`](https://github.com/kubernetes/api/blob/48bd8381a38a486f8b3cdf28cf7334a45b182f2e/certificates/v1/types.go#L41)
+resource graduated to `certificates.k8s.io/v1`. This makes the concept of
+requesting, signing, and consuming certificates in Kubernetes first class
+objects. cert-manager is well placed to serve this resource type as it becomes
+more popular in the wider community, whilst preserving all the features and
+extensions that cert-manager has to offer.
+
+
+## Motivation
```

Comment:
Feedback from call: let's expand this out to make it as clear as possible as to _why_ we want to adopt this type (so we are all on the same page :))

Reply:
The mains whys as I see it:

- Using an upstream API means there isn't a dependency on installing cert-manager's CRDs and its webhook before CSRs can be created by workloads, even if they are intended to be signed by cert manager later.
- Similar to the above point projects such as Istio would rather adopt the CSR API. CockroachDB already includes an example installation that signs database node certificates with the CSR API
- Avoid fragmentation in the wider ecosystem. We want cert-manager to be the CNCF standard, so not supporting an internal certificate API would be silly.